torch.nn.utils.clip_grad_norm_: remove device syncs

mautier opened the pull request:
Fixes #60691

### Changes

Per the discussion in the above issue, this PR makes 2 changes:
1. When `error_if_nonfinite=False`, the NaN/Inf checks are truly skipped, and no device synchronization occurs.
    - Additionally, when performing the checks, the 2 results are combined with `torch.logical_or` to incur only a single sync (instead of 2 in the happy/finite path).
2. The `clip_coef` conditional is removed, in favor of a call to `clamp(..., max=1.0)` and an unconditional multiplication.

### Testing

- The existing unit tests for `clip_grad_norm_` pass.
- I have manually profiled the example program from #60691, and verified that:
    - No synchronizations occur when using `error_if_nonfinite=False`.
    - A single synchronization occurs when using `error_if_nonfinite=True`.

mautier committed:
nn.utils.clip_grad_norm_: truly skip nan/inf checks when caller opts out

Previously, when the caller opted out of nan/inf checks on the gradient
norm, the checks would still be run in order to produce a warning.

Unfortunately doing so incurs a synchronization cost when the gradients
live on, say, a CUDA device, as the CPU-side control flow depends on the
result of CUDA-side computations.

This commit removes the warning codepath; when the user opts out of
finite-ness checks (`error_if_nonfinite=False`), the checks are skipped,
and there is no performance penalty.

Additionally, the 2 separate checks (`isnan()` and `isinf()`) are now
combined using `torch.logical_or`. This means that when the checks do
run on a non-CPU device, a single synchronization will be required
instead of 2. (the previous behavior allowed for short-circuiting, but
only in the NaN case, not in the happy path)

mautier committed:
nn.utils.clip_grad_norm_: avoid device sync with unconditional scaling

The `if clip_coef < 1:` conditional incurs a device synchronization when
the gradients are on a non-CPU device. This commit removes the
conditional in favor of a clamp-ing step and unconditional scaling.

mautier commented:
I briefly looked into the `pr/pytorch-linux-bionic-rocm4.2-py3.6` CI failure: `test_reduce_stress_cuda` in `test_c10d_gloo.py` is the failing test; as far as I can tell it is not related to `clip_grad_norm_` in any way, so let me know how to proceed!

Secondly, I'm not sure if the testing is did is sufficient: is there a way to assert in a test than 0 or 1 device sync is occuring during the call to `clip_grad_norm_`? Or is my manual check sufficient?

jbschlosser commented:
> I briefly looked into the `pr/pytorch-linux-bionic-rocm4.2-py3.6` CI failure: `test_reduce_stress_cuda` in `test_c10d_gloo.py` is the failing test; as far as I can tell it is not related to `clip_grad_norm_` in any way, so let me know how to proceed!

There are pre-existing failures for that test- we can ignore it for the purposes of this PR.

> Secondly, I'm not sure if the testing I did is sufficient: is there a way to assert in a test than 0 or 1 device sync is occuring during the call to `clip_grad_norm_`? Or is my manual check sufficient?

No way to assert on device syncs that I'm aware of - cc @ngimel for confirmation

jbschlosser reviewed (commented):
Changes on the python side lgtm - thanks for the update! Two things:
* Unfortunately, there's currently a C++ reimplementation of this function in `torch/csrc/api/include/torch/nn/utils/clip_grad.h` that should be updated as well
* I'm a bit curious about the performance implications of the unconditional multiplication on CPU

@JackCaoG Does it make sense to apply similar updates on the XLA side?

JackCaoG commented:
Thanks @jbschlosser , will take a look tmr.

JackCaoG commented:
@jbschlosser I think pt/xla can also benefit from this change, so I will update our patch as well. We are thinking about removing the patch for `clip_grad_norm_` for a while since `item` call is removed but we didn't benchmark the result yet..

JackCaoG commented:
Ah actually pt/xla already do the manual [`clapping`](https://github.com/pytorch/xla/blob/a010f8913c2eb86f664392a13e409a2ea11721c0/torch_xla/_patched_functions.py#L46) and check for [error_if_nonfinite](https://github.com/pytorch/xla/blob/a010f8913c2eb86f664392a13e409a2ea11721c0/torch_xla/_patched_functions.py#L40) already. Pt/xla should be fine.

jbschlosser commented:
@mautier Would you be willing to update the C++ version in `torch/csrc/api/include/torch/nn/utils/clip_grad.h` as well?

mautier commented:
@jbschlosser  Sorry for the radio silence; life things and a fried power supply have gotten in the way :upside_down_face: 

I have a patch for the C++ version that I'll push soon; unfortunately it appears that the C++ version has a slightly different API, in that it returns a `double` (instead of a scalar tensor like in python). So while I was able to eliminate the unwanted syncs and align the `error_if_nonfinite` semantics, there remains a final sync that cannot be removed without changing the function signature.

jbschlosser commented:
> @jbschlosser Sorry for the radio silence; life things and a fried power supply have gotten in the way 🙃
> 
> I have a patch for the C++ version that I'll push soon; unfortunately it appears that the C++ version has a slightly different API, in that it returns a `double` (instead of a scalar tensor like in python). So while I was able to eliminate the unwanted syncs and align the `error_if_nonfinite` semantics, there remains a final sync that cannot be removed without changing the function signature.

No worries! RIP your power supply :/

Thanks for checking out the C++ version as well! It's unfortunate that the API differences make that last sync unremovable, but it does makes sense. We want to keep the current API, but a comment about the discrepancy resulting in a final unremovable sync wouldn't hurt.

Commit:
torch::nn::utils::clip_grad_norm_: match py's error_if_nonfinite

Just like in the python version of this function, this commit removes
all nan/inf checks when `error_if_nonfinite = false`.

It also changes the computation of the norm to be based on standard
pytorch tensor operations (instead of more manual computations with
`std::max` and `std::pow`), in order to make it possible for those
computations to run directly on the device with no CPU synchronization.

Unfortunately, even then, since the C++ API returns the norm of the
gradients as a `double` (and not a scalar tensor), the implementation
must inevitably synchronize the CPU and device at the end.

Nonetheless, this function now synchronizes only once, as late as
possible, instead of many times (once per param).

albanD commented:
cc @ngimel

mautier commented on torch/csrc/api/include/torch/nn/utils/clip_grad.h:
```diff
@@ -23,43 +29,55 @@ inline double clip_grad_norm_(
+    total_norm_tensor = (norms.size() == 1) ? norms[0] : torch::max(torch::stack(norms));
   } else if (norm_type == 0) {
-    total_norm = static_cast<double>(params_with_grad.size());
+    total_norm_tensor = torch::full({}, static_cast<double>(params_with_grad.size()));
```
I'm not all that familiar with the C++ tensor creation APIs, is there a better way of creating a scalar tensor?

And on an unrelated note: order 0 norm is defined in [`torch.linalg.norm`](https://pytorch.org/docs/stable/generated/torch.linalg.norm.html#torch.linalg.norm) as the number of non-zero entries, but in this implementation it's just the number of parameters?

jbschlosser replied:
`torch::full()` is fine here imo.

Yeah, it does look like there's a discrepancy between here and the order 0 norm definition in `torch.linalg.norm`. I guess there's an implicit assumption that all grads are nonzero here? Regardless, we should leave as-is for backwards compatibility.

jbschlosser reviewed (approved):
Thanks for the awesome fix & commenting! LGTM

Mind rebasing to address merge conflicts?

mautier commented:
@jbschlosser Rebased on master and (force-)pushed. Let me know if you want me to squash the commits into one too (not sure if you merge or squash-merge PRs)!

mautier commented:
Rebased again to pick up 2 fixes that went into master since the last rebase; this should address the `quick-checks` and `shellcheck` lint failures.
